examples/stabilizer_code.py

```diff
@@ -0,0 +1,302 @@
+        # A1 = M[0:r, r : (n - k)]
+        A2 = M[0:r, (n - k) : n]
+        # B = M[0:r, n : (n + r)]
+        C1 = M[0:r, (n + r) : (2 * n - k)]
+        C2 = M[0:r, (2 * n - k) : (2 * n)]
+        # D = M[r : (n - k), n : (n + r)]
+        E = M[r : (n - k), (2 * n - k) : (2 * n)]
+
+        X = np.concatenate(
+            [
+                np.zeros((k, r), dtype=np.int8),
+                E.T,
+                np.eye(k, dtype=np.int8),
+                np.mod(E.T @ C1.T + C2.T, 2),
+                np.zeros((k, n - r), np.int8),
+            ],
+            axis=1,
+        )
+
+        Z = np.concatenate(
+            [
+                np.zeros((k, n), dtype=np.int8),
+                A2.T,
+                np.zeros((k, n - k - r), dtype=np.int8),
+                np.eye(k, dtype=np.int8),
+            ],
+            axis=1,
+        )
+
+        self.n: int = n
+        self.k: int = k
+        self.r: int = r
+        self.M: List[str] = _BuildByCode(M)
+        self.logical_Xs: List[str] = _BuildByCode(X)
+        self.logical_Zs: List[str] = _BuildByCode(Z)
+
+        self.syndromes_to_corrections = {}
+
+        for qid in range(self.n):
+            for op in allowed_errors:
+                syndrome = tuple(
+                    1 if self.M[r][qid] == 'I' or self.M[r][qid] == op else -1
+                    for r in range(self.n - self.k)
+                )
+                self.syndromes_to_corrections[syndrome] = (op, qid)
+
+    def encode(
+        self, additional_qubits: List[cirq.Qid], unencoded_qubits: List[cirq.Qid]
+    ) -> cirq.Circuit:
+        """
+        Creates a circuit that encodes the qubits using the code words.
+
+        Args:
+            additional_qubits: The list of self.n - self.k qubits needed to encode the qubit.
+            unencoded_qubits: The list of self.k qubits that contain the original message.
+
+        Returns:
+            A circuit where the self.n qubits are the encoded qubits.
+        """
+        assert len(additional_qubits) == self.n - self.k
+        assert len(unencoded_qubits) == self.k
+        qubits = additional_qubits + unencoded_qubits
+
+        circuit = cirq.Circuit()
+
+        # Equation 4.8:
+        # This follows the improvements of:
+        # https://cs269q.stanford.edu/projects2019/stabilizer_code_report_Y.pdf
+        for r, x in enumerate(self.logical_Xs):
+            for j in range(self.r, self.n - self.k):
+                if x[j] == 'X' or x[j] == 'Y':
+                    circuit.append(
+                        cirq.ControlledOperation(
+                            [unencoded_qubits[r]], cirq.X(additional_qubits[j])
+                        )
+                    )
+
+        gate_dict = {'X': cirq.X, 'Y': cirq.Y, 'Z': cirq.Z}
+
+        for r in range(self.r):
+            circuit.append(cirq.H(qubits[r]))
+
+            if self.M[r][r] == 'Y' or self.M[r][r] == 'Z':
+                circuit.append(cirq.S(qubits[r]))
```

Comment:
After reading the Thesis, its errata, and the reports, I think I fully understand this. Let's double-check if we understand on the same page:

1. Here `if self.M[r][r] == 'Y':` should be sufficient since in standard form it can never be `Z` in first `r` rows. Also, it has to be, otherwise, the argument of encoding is wrong.
2. The reason for adding `S` gate is `Y` gate we used is the complex format (i.e. to make it Hermitian). IIUC, the logic is this. First, consider the encoding `I+M_1` on initial state. It has following four cases: (ignore the phase factor)

> `(I+X@P_2...P_k)|0...0> = |0...0> + |1>|\psi>`
> `(I+Y@P_2...P_k)|0...0> = |0...0> + i|1>|\psi>`
> `(I+I@P_2...P_k)` and `(I+Z@P_2...P_k)` is not possible in standard form

The first case means we need `[1,1]` vector and controlled gates and in the second case we need `[1, i]` vector and controlled gates. Corresponding, it is the first column of `H` and the first column of `SH` respectively. 

Next is to build `I+M_l` and the trick is that all generators of stabilizer commute. Hence all implementation `I+M_l` is equivalent to apply in the first place, i.e. considering its effect on `|0...0>` only and then move this back to place. (Original thesis make this mismake not make entire gates back to place). So the argument repeats the same.

3. This implementation is limited to the case that all of the signs of the stabilizer generators (in standard form) are +1. It is possible that "Codes with other signs can be achieved via a tensor product Pauli operator." That is stated in Proposition 10.4 in Nielsen & Chuang's book. I am not sure hows easy makes it systematically

Reply:
Yes, that was indeed my understanding, but I have to re-read it every time. This is where I am the most iffy in my understanding, if I may be honest.

I took the liberty of adding comments that were largely inspired from what you wrote here. I hope that's OK.